machine/parser.go

```diff
@@ -27,67 +27,35 @@ func FromJSON(raw []byte) (*StateMachine, error) {
 	return &sm, err
 }
 
-type basicMachine struct {
-	Comment *string
-	StartAt *string
-}
-
-// from http://gregtrowbridge.com/golang-json-serialization-with-interfaces/
-func (sm *StateMachine) UnmarshalJSON(b []byte) error {
-	// First, deserialize StateMachine fields
-	var bsm basicMachine
-	err := json.Unmarshal(b, &bsm)
-	if err != nil {
-		return err
-	}
-	sm.Comment = bsm.Comment
-	sm.StartAt = bsm.StartAt
-
+func (sm *States) UnmarshalJSON(b []byte) error {
 	// States
-	var objMap map[string]*json.RawMessage
-	err = json.Unmarshal(b, &objMap)
+	var rawStates map[string]*json.RawMessage
+	err := json.Unmarshal(b, &rawStates)
+
 	if err != nil {
 		return err
 	}
 
-	var initStates map[string]*json.RawMessage
-
-	if val, ok := objMap["States"]; ok {
-		err = json.Unmarshal(*val, &initStates)
+	newStates := States{}
+	for name, raw := range rawStates {
+		states, err := unmarshallState(name, raw)
 		if err != nil {
 			return err
 		}
 
-		sm.States = map[string]state.State{}
-		err = unmarshallStates(sm, initStates)
-		if err != nil {
-			return err
+		for _, s := range states {
+			newStates[*s.Name()] = s
 		}
 	}
 
+	*sm = newStates
 	return nil
 }
 
 type stateType struct {
 	Type string
 }
 
-func unmarshallStates(sm *StateMachine, initStates map[string]*json.RawMessage) error {
-	for name, raw_json := range initStates {
-		states, err := unmarshallState(name, raw_json)
-		if err != nil {
-			return err
-		}
-		for _, state := range states {
-			if err := sm.AddState(state); err != nil {
-				return err
-			}
-		}
-	}
-
-	return nil
-}
-
 func unmarshallState(name string, raw_json *json.RawMessage) ([]state.State, error) {
```

Comment:
nit: I think that you should put unmarshallState on the state class. Then (in theory), all you need to do in this file is:

```
var newStates States{}
err := json.Unmarshal(b, &newStates)
if err ...
*sm = newStates
return nil
```

Reply:
`State` is an interface so the problem is that you don't know what concrete struct is needed without looking into the object for the `Type`. This means that you have to do like 2 pass unmarshalling, first to a temporary struct, then to a concrete one. 

Also, `err := json.Unmarshal(b, &newStates)` would cause a recursion because that is the method being implemented here.